rust/apps/zcash/src/pczt/check.rs

```diff
@@ -0,0 +1,343 @@
+    keys::{FullViewingKey, Scope},
+    note::{Memo, Note, Nullifier, Rho},
+    note_ext::{calculate_note_commitment, calculate_nullifier},
+    Address,
+};
+use parse::decode_output_enc_ciphertext;
+use zcash_vendor::{
+    bip32::ChildNumber,
+    pczt,
+    ripemd::Ripemd160,
+    sha2::{Digest, Sha256},
+    zcash_primitives::legacy::{keys::AccountPubKey, Script, TransparentAddress},
+};
+
+pub fn check_pczt(
+    seed_fingerprint: &[u8; 32],
+    ufvk: &UnifiedFullViewingKey,
+    pczt: &Pczt,
+) -> Result<(), ZcashError> {
+    let xpub = ufvk.transparent().ok_or(ZcashError::InvalidDataError(
+        "transparent xpub is not present".to_string(),
+    ))?;
+    let orchard = ufvk.orchard().ok_or(ZcashError::InvalidDataError(
+        "orchard fvk is not present".to_string(),
+    ))?;
+    check_orchard(&seed_fingerprint, &orchard, &pczt.orchard())?;
+    check_transparent(&seed_fingerprint, &xpub, &pczt.transparent())?;
+    Ok(())
+}
+
+fn check_transparent(
+    seed_fingerprint: &[u8; 32],
+    xpub: &AccountPubKey,
+    bundle: &pczt::transparent::Bundle,
+) -> Result<(), ZcashError> {
+    bundle.inputs().iter().try_for_each(|input| {
+        check_transparent_input(seed_fingerprint, xpub, input)?;
+        Ok(())
+    })?;
+    bundle.outputs().iter().try_for_each(|output| {
+        check_transparent_output(seed_fingerprint, xpub, output)?;
+        Ok(())
+    })?;
+    Ok(())
+}
+
+fn check_transparent_input(
+    seed_fingerprint: &[u8; 32],
+    xpub: &AccountPubKey,
+    input: &pczt::transparent::Input,
+) -> Result<(), ZcashError> {
+    let script = Script(input.script_pubkey().clone());
+    match script.address() {
+        Some(TransparentAddress::PublicKeyHash(hash)) => {
+            let pubkey = input.bip32_derivation().keys().find(|pubkey| {
+                return hash[..] == Ripemd160::digest(Sha256::digest(pubkey))[..];
+            });
+            match pubkey {
+                Some(pubkey) => {
+                    match input.bip32_derivation().get(pubkey) {
+                        Some(bip32_derivation) => {
+                            if seed_fingerprint == &bip32_derivation.seed_fingerprint {
+                                //verify public key
+                                let xpub = xpub.to_inner();
+                                let target = bip32_derivation.derivation_path.iter().try_fold(
+                                    xpub,
+                                    |acc, n| {
+                                        acc.derive_child(ChildNumber::from(*n)).map_err(|_| {
+                                            ZcashError::InvalidPczt(
+                                                "transparent input bip32 derivation path invalid"
+                                                    .to_string(),
+                                            )
+                                        })
+                                    },
+                                )?;
```

Comment:
Bug: `xpub` is obtained earlier as `ufvk.transparent()`, which means it is already derived at the path `M/44'/133'/account_index'`. However, here the code is applying the transparent inputs entire `bip32_derivation` path to `xpub`, which results in the key at `M/44'/133'/account_index'/44'/133'/account_index'/scope/address_index` (which itself doesn't make sense as you'd be trying to do hardened derivation from a public key, so you're probably just getting an error here immediately).

Instead, the code should check that `bip32_derivation.derivation_path` starts with `44'/133'/account_index'`, and then only derive from `xpub` with the remaining suffix (which should be entirely non-hardened for a valid input, which you'll detect via the existing error check), to obtain (in the usual Zashi tx flow) `M/44'/133'/account_index'/scope/address_index`.

Reply:
This is fixed now that the firmware is using `AccountPubKey::derive_pubkey_at_bip32_path`.